File: First Year/Second Semester/Graph Algorithms/Assignment1CPP/FileManager/FileManager.cpp
//
// Created by Emanuel on 28.03.2021.
//

#include "FileManager.h"

#include <iostream>
#include <fstream>

DirectedGraph FileManager::read(const std::string &file_path) const {
    std::ifstream fin;
    fin.open(file_path);
    int vertices {0}, edges {0}, edge_x {0}, edge_y {0}, cost {0};

    fin >> vertices >> edges;
    DirectedGraph graph {vertices};

    for (size_t i = 0; i < edges; ++i) {
        fin >> edge_x >> edge_y >> cost;
        graph.addEdge(edge_x, edge_y, cost);
    }

    fin.close();
    return graph;
}

void FileManager::write(DirectedGraph graph, const std::string &file_path) {
    std::ofstream fout;
    fout.open(file_path);
    std::string file_data = "";

    file_data += std::to_string(graph.getVerticesNo()) + " " + std::to_string(graph.getEdgesNo()) + "\n";
    std::vector<int> isolated_nodes;

    auto vertices_it = graph.parseVertices();
    while (vertices_it.isValid()) {
        int vertex = vertices_it.getCurrent();
        if (!graph.getOutDegree(vertex))
            isolated_nodes.push_back(vertex);
        auto out_it = graph.parseVerticesOut(vertex);
        while (out_it.isValid()) {
            int neighbour = out_it.getCurrent();
            file_data += std::to_string(vertex) + " " + std::to_string(neighbour) + " " + std::to_string(graph.getCost(vertex, neighbour)) + "\n";
            out_it.next();
        }
        vertices_it.next();
    }

    for (const auto& node : isolated_nodes)
        file_data += std::to_string(node) + "\n";
    fout << file_data;

    fout.close();
}

File: First Year/Second Semester/Graph Algorithms/Assignment5/FileManager/FileManager.py
from Domain.DirectedGraph import DirectedGraph


class FileManager:
    @staticmethod
    def read(file_path, type=DirectedGraph):
        """
            Reads graph data from a file
            :param type: the type of the read graph
            :param file_path: the path of the file
            :return: the graph read from the file
        """
        with open(file_path, 'r') as text_file:
            vertices, edges = [int(data) for data in next(text_file).split()]  # read the first line
            graph = type(vertices)
            for edge in range(edges):  # read the edges
                edge_x, edge_y, cost = [int(data) for data in next(text_file).split()]  # read the next line
                graph.addEdge(edge_x, edge_y, cost)
        return graph

    @staticmethod
    def write(graph, file_path):
        """
            Writes graph data to a file
            :param graph: the graph to be written into a file
            :param file_path: the path to the file
        """
        vertices_no, edges_no = graph.getVerticesNo(), graph.getEdgesNo()
        file_data = str(vertices_no) + " " + str(edges_no) + "\n"

        isolated_nodes = []
        for vertex in graph.parseVertices():
            if not graph.getOutDegree(vertex):
                isolated_nodes.append(vertex)
            for neighbour in graph.parseVerticesOut(vertex):
                file_data += str(vertex) + " " + str(neighbour) + " " + str(graph.getCost(vertex, neighbour)) + "\n"
        for node in isolated_nodes:
            file_data += str(node) + "\n"

        with open(file_path, "w") as text_file:
            text_file.write(file_data)
